src/uu/cp/src/cp.rs

```diff
@@ -710,43 +774,45 @@ impl Options {
         };
 
         // Parse attributes to preserve
-        let mut preserve_attributes: Vec<Attribute> = if matches.contains_id(options::PRESERVE) {
+        let attributes: Attributes = if matches.contains_id(options::PRESERVE) {
             match matches.get_many::<String>(options::PRESERVE) {
-                None => DEFAULT_ATTRIBUTES.to_vec(),
+                None => Attributes::default(),
                 Some(attribute_strs) => {
-                    let mut attributes = Vec::new();
+                    let mut attributes: Attributes = Attributes::none();
+                    let mut attributes_empty = true;
                     for attribute_str in attribute_strs {
+                        attributes_empty = false;
                         if attribute_str == "all" {
-                            attributes = add_all_attributes();
+                            attributes.max(Attributes::all());
                             break;
                         } else {
-                            attributes.push(Attribute::from_str(attribute_str)?);
+                            attributes.try_set_from_string(
+                                attribute_str,
+                                Attributes::default_explicit(),
+                            )?;
```

Comment:
If this is the only `Attributes` you pass to `try_set_from_string` I think it's worth simplifying and always make `try_set_from_string` set attributes to `Preserve::Yes { required: true }`.

Reply:
I.e.
```rust
fn try_set_from_string(&mut self, value: &str) -> Result<(), Error> {
    let yes = Preserve::Yes { required: yes };
    match &*value.to_lowercase() {
        "mode" => self.mode = yes,
        #[cfg(unix)]
        "ownership" => self.ownership = yes,
        "timestamps" => self.timestamps = yes,
        #[cfg(feature = "feat_selinux")]
        "context" => self.context = yes,
        "links" => self.links = yes,
        "xattr" => self.xattr = yes,
        _ => {
            return Err(Error::InvalidArgument(format!(
                "invalid attribute {}",
                value.quote()
            )));
        }
    };
    Ok(())
}
```

Reply:
Oh and `"context"` should probably be recognized but emit a warning instead of failing if `feat_selinux` is not enabled.

Reply:
> Oh and `"context"` should probably be recognized but emit a warning instead of failing if `feat_selinux` is not enabled.

Thinking about this right now, should I then throw a new error type? Maybe `FeatureDisabled`? Maybe something like this:
```rust
"context" =>
    {
        #[cfg(feature = "feat_selinux")]
        {
            self.context = self.context.max(preserve_yes_required);
        }
        #[cfg(not(feature = "feat_selinux"))]
        {
            return Err(Error::FeatureDisabled)
        }
    }
,
```

Reply:
Hm, then it makes sense to throw an error in `handle_preserve(&attributes.context, || -> CopyResult<()> {...` section, it handles errors as warnings if not required already :)

Reply:
Ah, never mind, it just ignores silently if it's not required. Hm, would it be better to print the error message in the `handle_preserve` if the attribute in not required (and throw like we have it now otherwise)? I think this should be a more helpful behavior.

Reply:
Hm, this breaks `test_cp_parents_2_dirs` test, since it enables `--archive` and that causes it to print warning for the context option, which is not desired if I judge by the test?

Reply:
How about having the default attributes for the all case have the following conditioning for the context, if the cfg feature feat_selinux is on, then set context to `Preserve::Yes { required: false }`, otherwise to `Preserve::No,`, then only if the user explicitly specifies the `context` it will warn.

Reply:
That makes sense to me! About handling it in `handle_preserve`, wouldn't it then print the warning for each file? That seems a bit much.

Reply:
How about now?